### PR TITLE
feat: support `i64::checked_mod`

### LINKED ISSUE #1000: Implement `checked_mod` for `i64`
Start at `OpEmitter::checked_mod` and add `OpEmmitter::checked_mod_i64` that calls the new intrinsic (add it) at `codegen/masm/intrinsics/i64.masm`.

### PR DESCRIPTION
**Pull request by mooori**
- Adds intrinsic `i64::checked_mod`
  - semantics correspond to Rust's [`checked_rem`](https://doc.rust-lang.org/std/primitive.i32.html#method.checked_rem), see #1174 for the reasoning
- Adds codegen for `checked_mod_i64`

Closes #1000 and unlocks `i64::overflowing_rem`, `i64::checked_rem`.

### CODE REVIEW

**Comment by mooori** [comment]
I think the `lit` CI failure is unrelated to the changes in this PR:

- The `lit` suite has been failing silently for a while and the fixes from #1175 are not yet merged.
- The `litcheck` exit code issue has already been [fixed and published](https://github.com/0xMiden/compiler/pull/1175#pullrequestreview-4519193550)

So I suppose one of the pre-existing issues is now making the `lit` CI job red.

File: tests/integration/src/end_to_end/intrinsics/i64_arithmetic.rs [hunk 1]
```diff
@@ -363,6 +363,30 @@ fn i64_checked_div() {
...
+    # Stack: [r_lo, r_hi]
+"#;
+    test_i64_intrinsic(
+        proc_body,
+        NumericStrategy::<i64>::rem_signed_checked(),
+        binary_i64op_inputs_to_stack,
+        |(a, b): &(i64, i64)| {
+            if *b == 0 {
+                Err(TrapExpectation::DivideByZero)
+            } else if *a == i64::MIN && *b == -1 {
```
Comment (greenhat) [inline tests/integration/src/end_to_end/intrinsics/i64_arithmetic.rs:380]: WebAssembly signed remainder is only undefined for a zero divisor; `i64.rem_s(i64::MIN, -1)` should return `0`, not trap ([WebAssembly numerics, `irem_s`](https://webassembly.github.io/spec/core/exec/numerics.html#op-irem)).

**Review by greenhat — CHANGES_REQUESTED** [review]
Looking good overall! Please check my note.
Also, IIRC we have `arith.smod`. Should we use it here?

File: codegen/masm/intrinsics/i64.masm [hunk 2]
```diff
@@ -253,6 +253,40 @@ pub proc checked_div # [b_lo, b_hi, a_lo, a_hi]
     end
 end
 
+# Computes `a % b`, asserting that both inputs are valid i64 values (u32 limbs).
+#
+# Execution traps on division by zero, and on `i64::MIN % -1` (the underlying division overflows).
+pub proc checked_mod # [b_lo, b_hi, a_lo, a_hi]
```
Comment (bitwalker) [inline codegen/masm/intrinsics/i64.masm:259]: We should add type signatures to all new intrinsics

**Review by bitwalker — COMMENTED** [review]
Looks good! Aside from @greenhat's notes, I just had one nit, otherwise 👍

File: tests/integration/src/end_to_end/intrinsics/i64_arithmetic.rs (hunk 1 above)
Comment (mooori) [reply tests/integration/src/end_to_end/intrinsics/i64_arithmetic.rs:380]: > i64.rem_s(i64::MIN, -1) should return 0, not trap

I think this means we have a bug in the lowering of Wasm's `i64.rem_s` which tests do not catch:

- Wasm `i64.rem_s` is probably compiled to `i64::checked_mod` intrinsic
- For `(i64::MIN, -1)` Wasm returns 0 but the intrinsic traps

The arith tests `i64::checked_rem` and `i64::overflowing_rem` do hit the intrinsic `i64::checked_mod`. Though when compiling `i64::checked_rem`, `i64::overflowing_rem` from Rust to Wasm, the generated Wasm handles edge cases explicitly, thus hiding the bug.

My plan is:

- Add test(s) to check whether the `i64.rem_s` lowering is fine or not
- Fix it if needed. Maybe by adding intrinsic `i64::wrapping_mod` and lowering `i64.rem_s` to that.

The same issue should then also exist for `i32.rem_s` and the `i32::checked_mod` intrinsic.

Comment (greenhat) [reply tests/integration/src/end_to_end/intrinsics/i64_arithmetic.rs:380]: Sounds good!

**Comment by mooori** [comment]
Reverted to draft to work on #1206 first.
